### PR TITLE
media-libs/libgroove: new ebuild

### PR DESCRIPTION
**Pull request by diogocp**
A streaming audio processing library ([andrewrk/libgroove](https://github.com/andrewrk/libgroove)).
Gentoo-Bug: [558190](https://bugs.gentoo.org/show_bug.cgi?id=558190)

### CODE REVIEW

File: media-libs/libgroove/libgroove-4.3.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+DESCRIPTION="Streaming audio processing library."
+HOMEPAGE="https://github.com/andrewrk/libgroove"
+SRC_URI="https://github.com/andrewrk/libgroove/archive/${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64"
+IUSE="+chromaprint +loudness +sound"
+
+DEPEND="media-video/ffmpeg
```
Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: * no USE flags needed for ffmpeg? sure?
* does it not work with `media-video/libav`?

File: media-libs/libgroove/Manifest [hunk 2]
```diff
@@ -0,0 +1,3 @@
+DIST libgroove-4.3.0.tar.gz 48857 SHA256 76f68896f078a9613d420339ef887ca8293884ad2cd0fbc031d89a6af2993636 SHA512 30d75a3b1b3efa9b33687bb146648b66f8f4c718551c0db2c24c0259f09cf355a7c7bbac40d60d7a97bbbd7f83401d09464c6ac2b18cdfcd6198220cefc8a705 WHIRLPOOL 4040769723e7891b285318d538cf628203289aca7f7a655855716fe450f58712cd64b2fdc2abbd316bef6f80bf6b65ae6620b9ac339b3ebd5c7e091a36129040
+EBUILD libgroove-4.3.0.ebuild 794 SHA256 0074fb7f431bc884a75449c5927b2d96a1d6b176fa6ceeb8273fe0f5c6720737 SHA512 cbffbb8b868a43989056199c31c0a3a2cb6d1da51b96d920b60ba8590409d32fcde635d00031dfc7a6f6089898f9c056e8bdae41e088c8ac5094515de3ca3bc6 WHIRLPOOL 7bc70a6373f690b1b3b8886ace7c099c6fc42b100202211dad733b8d5a36376b6ac6e1ca4ec7cf65d6207280059d7dd64b41d9a71d759858c3a99d24ebe8a31b
```
Comment (hasufell) [inline media-libs/libgroove/Manifest]: wut? :o 
why is there an ebuild hash?

File: media-libs/libgroove/metadata.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
```
Comment (hasufell) [inline media-libs/libgroove/metadata.xml]: missing maintainer information here, e.g.
```xml
	<herd>proxy-maintainers</herd>
	<maintainer>
		<email>YOUR@EMAIL.COM</email>
		<name>YOUR NAME</name>
		<description>Proxy maintainer. Assign bugs to him.</description>
	</maintainer>
```

**Comment by hasufell** [comment]
You should reference this bug report https://bugs.gentoo.org/show_bug.cgi?id=558190 somewhere in your commit message. Since there is no consensus on how to do that, just do something (e.g. `wrt bug 558190` or in the multiline commit description `Gentoo-Bug: 558190` or whatever)

File: media-libs/libgroove/Manifest (hunk 2 above)
Comment (diogocp) [inline media-libs/libgroove/Manifest]: Shouldn't there be? I just ran `ebuild libgroove-4.3.0.ebuild digest`.

Comment (hasufell) [inline media-libs/libgroove/Manifest]: run `repoman manifest`

File: media-libs/libgroove/metadata.xml (hunk 3 above)
Comment (diogocp) [inline media-libs/libgroove/metadata.xml]: Should I put my own info? Usually a dev does it.

Comment (hasufell) [inline media-libs/libgroove/metadata.xml]: A dev does what? There is no maintainer for it currently. I guess you want to maintain it.

Comment (diogocp) [inline media-libs/libgroove/metadata.xml]: Usually I open a bug with a new ebuild and the dev who commits it becomes the maintainer. If it's different now, no problem, I'm just not used to the new workflow.

File: media-libs/libgroove/Manifest (hunk 2 above)
Comment (diogocp) [inline media-libs/libgroove/Manifest]: `repoman manifest` also outputs the ebuild hash.

File: media-libs/libgroove/metadata.xml (hunk 3 above)
Comment (hasufell) [inline media-libs/libgroove/metadata.xml]: > Usually I open a bug with a new ebuild and the dev who commits it becomes the maintainer.

selfish gentoo devs!

File: media-libs/libgroove/Manifest (hunk 2 above)
Comment (hasufell) [inline media-libs/libgroove/Manifest]: @mgorny do you have an idea?

Comment (mgorny) [inline media-libs/libgroove/Manifest]: Do you run `repoman manifest` in the local clone of Gentoo repository, with `metadata/layout.conf` intact? What version of Porage do you have installed?

Comment (diogocp) [inline media-libs/libgroove/Manifest]: That was it, I was running it in my local overlay.

File: media-libs/libgroove/libgroove-4.3.0.ebuild (hunk 1 above)
Comment (diogocp) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: * Upstream has some suggested flags, but from my tests they are all optional runtime dependencies.
* Support for libav was recently removed, but it looks like this version still supports it. I think the default should be ffmpeg, to minimize problems when upgrading to the next version, unless @andrewrk thinks libav is better for 4.3.0.

Comment (andrewrk) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: ffmpeg should be fine for 4.3.0. The next release of libgroove will officially only support ffmpeg.

**Comment by diogocp** [comment]
OK, I think that's all. Please take another look.

**Comment by hasufell** [comment]
The build system unfortunately uses `-Werror`. We will have to fix/patch that, also see https://devmanual.gentoo.org/ebuild-writing/common-mistakes/index.html under `-Werror compiler flag not removed`

**Comment by hasufell** [comment]
We should probably provide a `static-libs` USE flag and `rm` the `foo.a` files when it is turned off in `src_install`.

File: media-libs/libgroove/libgroove-4.3.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+EAPI=5
+
+inherit cmake-utils
+
+DESCRIPTION="Streaming audio processing library."
+HOMEPAGE="https://github.com/andrewrk/libgroove"
+SRC_URI="https://github.com/andrewrk/libgroove/archive/${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
```
Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: We should probably provide a SUBSLOT, in this case `SLOT="0/4"` also see https://wiki.gentoo.org/wiki/Sub-slots_and_Slot-Operators

**Comment by andrewrk** [comment]
> The build system unfortunately uses -Werror. We will have to fix/patch that, also see https://devmanual.gentoo.org/ebuild-writing/common-mistakes/index.html under -Werror compiler flag not removed
> 
> Here are the offending lines https://github.com/andrewrk/libgroove/blob/master/CMakeLists.txt#L126 ...which also reveal -g in the compiler flags. That has to go out as well.

Upstream maintainer here. My goal is to be friendly to package maintainers like you. I believe that cmake will handle the `-g` flag correctly depending on the build type, so I'm happy to simply remove that upstream.

As for `-Werror`, I do want that on for development, but is there a way I could prevent gentoo from having to patch it away? Perhaps if I enabled `-Werror` only for debug builds and not for release builds?

**Comment by hasufell** [comment]
> Perhaps if I enabled -Werror only for debug builds and not for release builds?

Gentoo hacks it's own cmake build type in (although we sometimes fall back to release build type if that causes trouble), so yes, that would probably work.

File: media-libs/libgroove/libgroove-4.3.0.ebuild (hunk 4 above)
Comment (diogocp) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: I see that it installs both `libgroove.so.4` and `libgroove.so.4.3.0`. How do we decide between `0/4` and `0/4.3.0`?

**Comment by andrewrk** [comment]
ok. I made this change upstream: https://github.com/andrewrk/libgroove/commit/0b619fa3b02ab55593f148bfd867aec2c34f66c1

If it would really help to make a new release I could do that. Otherwise I'd rather wait until some other stuff I'm presently working on is done to make a new release.

**Comment by hasufell** [comment]
should be fine to use the patch https://github.com/andrewrk/libgroove/commit/0b619fa3b02ab55593f148bfd867aec2c34f66c1.patch

**Comment by andrewrk** [comment]
don't forget the second commit there. Sorry, I missed the -Werror and -pedantic in the example flags at first.

File: media-libs/libgroove/libgroove-4.3.0.ebuild (hunk 4 above)
Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]:
```
# objdump -p /usr/lib64/libgrooveplayer.so.4.3.0 | grep SONAME  SONAME
libgrooveplayer.so.4
```
This reflects ABI compatibility.

Comment (diogocp) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: Interesting, thanks.

**Comment by hasufell** [comment]
https://github.com/andrewrk/libgroove/compare/44aa5db96cfe45e56fa1f7d762fb032ca178a35c...master.diff

**Comment by diogocp** [comment]
I think this is the one: https://github.com/andrewrk/libgroove/compare/44aa5db96cfe45e56fa1f7d762fb032ca178a35c...master.patch

**Comment by diogocp** [comment]
They're the same thing :dizzy_face:

**Comment by diogocp** [comment]
I had to make another patch because that one failed. :/

File: media-libs/libgroove/libgroove-4.3.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		$(cmake-utils_use_disable chromaprint FINGERPRINTER)
+		$(cmake-utils_use_disable loudness LOUDNESS)
+		$(cmake-utils_use_disable sound PLAYER)
+	)
+	cmake-utils_src_configure
+}
+
+src_install() {
+	cmake-utils_src_install
+	use static-libs || rm "${D}"/usr/lib/*.a
```
Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: nah, this needs improvement, it should be
```
use static-libs || { rm "${ED%/}"/usr/$(get_libdir)/*.a || die "failed to remove static libs!" ;}
```
and then we have to `inherit multilib cmake-utils`

File: media-libs/libgroove/files/libgroove-4.3.0_cflags.patch [hunk 6]
```diff
@@ -0,0 +1,18 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
```
Comment (hasufell) [inline media-libs/libgroove/files/libgroove-4.3.0_cflags.patch]: reference the upstream commit messages with an URL here and give a short summary like "remove -Werror from compiler flags"

File: media-libs/libgroove/libgroove-4.3.0.ebuild (hunk 5 above)
Comment (diogocp) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: Or inherit cmake-utils multilib? I noticed [this discussion](https://github.com/gentoo/gentoo/pull/1#discussion-diff-36746284) before, but have not seen any guidelines.

Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: I personally order them by priority, to minimalize the chance that something weird happens when exported phases change. So something like `inherit <utility.eclass> <heavy-phase-export.eclass>`. But I don't really care here, as long as the end result works.

Comment (diogocp) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: It's failing because get_libdir returns lib64 but they are installed in /usr/lib.

Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: That's a bug then that needs to be fixed/patched.

Comment (hasufell) [inline media-libs/libgroove/libgroove-4.3.0.ebuild]: https://github.com/andrewrk/libgroove/pull/100

**Comment by hasufell** [comment]
excellent, I think this is good to merge, almost..

@diogocp could you rebase your `libgroove` against latest gentoo master?
Something like:
```sh
git pull --rebase=preserve https://anongit.gentoo.org/git/repo/gentoo.git master
git push -f origin libgroove
```

**Comment by diogocp** [comment]
Done

**Comment by diogocp** [comment]
Thanks!

**Pull request merged**